[AMDGPU] SIPeepholeSDWA: Handle V_CNDMASK_B32_e64

frederik-h opened the pull request:
The VOP3 form of the V_CNDMASK_B32 instruction takes a carry-in
operand. The conversion to SDWA implies a conversion to VOP2 form
which reads from VCC instead.

Convert V_CNDMASK_B32_e64 instructions that might be converted to SDWA
to V_CNDMASK_B32_e32 first and introduce a copy of the carry-in operand to VCC.

Closes #133431.

frederik-h committed:
[AMDGPU] SIPeepholeSDWA: Reject V_CNDMASK_B32_e64 instead of V_CNDMASK_B32_e32

The problem with V_CNDMASK_B32_e64 (i.e. that conversion to the VOP2
SDWA form introduces an implicit VCC use) hinted at by the comment
does not exist with V_CNDMASK_B32_e32. Hence the latter should already
be acceptable for conversion to SDWA without further ado.

frederik-h committed:
[AMDGPU] SIPeepholeSDWA: Handle V_CNDMASK_B32_e64

The VOP3 form of the V_CNDMASK_B32 instruction takes a carry-in
operand. The conversion to SDWA implies a conversion to VOP2 form
which reads from VCC instead.

Convert V_CNDMASK_B32_e64 instructions that might be converted to SDWA
to V_CNDMASK_B32_e32 first and either change the instruction that
defines the carry-in operand to write to VCC if this is possible or
introduce a write of the carry-in operand to VCC.

arsenm commented on llvm/lib/Target/AMDGPU/SIPeepholeSDWA.cpp:
```diff
@@ -1061,6 +1062,79 @@ void SIPeepholeSDWA::pseudoOpConvertToVOP2(MachineInstr &MI,
+  assert(MI.getOpcode() == AMDGPU::V_CNDMASK_B32_e64);
+
+  MCRegister Vcc = TRI->getVCC();
+  // FIXME Conversion introduces implicit vcc_hi use
+  if (Vcc == AMDGPU::VCC_LO)
+    return;
```
Need to fix the fixme, this is just dropping support for wave32. The way we handle wave32 is not good, the definitions are wave64 and we have to hack on the instruction after creation to adjust it to wave32. e.g. see fixImplicitOperands

frederik-h replied:
Thanks for the pointer to the function. I had to add the fix to the function that creates the sdwa instruction instead of the new conversion function.

arsenm commented on llvm/lib/Target/AMDGPU/SIPeepholeSDWA.cpp:
```diff
@@ -1061,6 +1062,79 @@ void SIPeepholeSDWA::pseudoOpConvertToVOP2(MachineInstr &MI,
+
+  // Make sure VCC or its subregs are dead before MI.
+  MachineBasicBlock &MBB = *MI.getParent();
+  auto Liveness = MBB.computeRegisterLiveness(TRI, Vcc, MI, 100);
```
No auto, and do this after all other checks below. Also can just use the default threshold, 100 is a lot

arsenm commented on llvm/lib/Target/AMDGPU/SIPeepholeSDWA.cpp:
```diff
@@ -1061,6 +1062,79 @@ void SIPeepholeSDWA::pseudoOpConvertToVOP2(MachineInstr &MI,
+    CarryDef.substituteRegister(CarryIn.getReg(), Vcc, 0, *TRI);
+    CarryDef.moveBefore(&MI);
+  } else {
+    // Add write: VCC[lanedId] <- (CarryIn[laneId] == 1)
+    const TargetRegisterClass *Class =
+        TRI->getRegClassForOperandReg(*MRI, CarryIn);
+    unsigned RegSize = Class->MC->getSizeInBits();
+    BuildMI(MBB, MI, MI.getDebugLoc(), TII->get(getVCmpEqOpcode(RegSize)))
+        .addReg(Vcc, RegState::Define)
+        .addImm(1)
+        .add(CarryIn);
```
You can't introduce a new compare, at worst you should introduce a copy

arsenm commented on llvm/lib/Target/AMDGPU/SIPeepholeSDWA.cpp:
```diff
@@ -1061,6 +1062,79 @@ void SIPeepholeSDWA::pseudoOpConvertToVOP2(MachineInstr &MI,
+  // Change destination of compare instruction to VCC
+  // or copy to VCC if carry-in is not a compare inst.
+  Register CarryReg = CarryIn.getReg();
+  MachineInstr &CarryDef = *MRI->getVRegDef(CarryReg);
```
This will fail if CarryIn is undef

arsenm commented on llvm/lib/Target/AMDGPU/SIPeepholeSDWA.cpp:
```diff
@@ -1061,6 +1062,79 @@ void SIPeepholeSDWA::pseudoOpConvertToVOP2(MachineInstr &MI,
+  if (CarryDef.isCompare() && TII->isVOP3(CarryDef) &&
+      MRI->hasOneUse(CarryIn.getReg())) {
+    CarryDef.substituteRegister(CarryIn.getReg(), Vcc, 0, *TRI);
+    CarryDef.moveBefore(&MI);
```
It might not be safe to move based on the other operands, but you don't have any checks for that. Should just not move the instruction, it's beyond the scope of what this pass should deal with

arsenm commented on llvm/lib/Target/AMDGPU/SIPeepholeSDWA.cpp:
```diff
@@ -1061,6 +1062,79 @@ void SIPeepholeSDWA::pseudoOpConvertToVOP2(MachineInstr &MI,
+                       .add(*TII->getNamedOperand(MI, AMDGPU::OpName::src0))
+                       .add(*TII->getNamedOperand(MI, AMDGPU::OpName::src1))
+                       .setMIFlags(MI.getFlags());
+  LLVM_DEBUG(dbgs() << "Converted to VOP2: " << *Converted << '\n');
```
```suggestion
  LLVM_DEBUG(dbgs() << "Converted to VOP2: " << *Converted);
```
MachineInstr << already includes a newline

arsenm commented on llvm/test/CodeGen/AMDGPU/sdwa-peephole-vcnd_mask-1.mir:
```diff
@@ -0,0 +1,48 @@
+    ; gfx9-NEXT: [[V_PERM_B32_e64_:%[0-9]+]]:vgpr_32 = V_PERM_B32_e64 killed [[V_CNDMASK_B32_sdwa]], killed [[V_CNDMASK_B32_e64_]], killed [[S_MOV_B32_]], implicit $exec
+    ; gfx9-NEXT: $vgpr0 = COPY [[V_PERM_B32_e64_]]
+    ; gfx9-NEXT: SI_RETURN implicit $vgpr0
+    %10:vgpr_32 = COPY $vgpr2
```
Compact the register numbers

arsenm commented on llvm/test/CodeGen/AMDGPU/sdwa-peephole-vcnd_mask-1.mir:
```diff
@@ -0,0 +1,48 @@
+    %17:vgpr_32 = V_CNDMASK_B32_e64 0, 0, 0, %10, killed %14, implicit $exec
+    %20:vgpr_32 = V_LSHRREV_B32_e64 16, %10, implicit $exec
+    %22:vgpr_32 = V_CNDMASK_B32_e64 0, 0, 0, %20, killed %12, implicit $exec
+    %24:sreg_32 = S_MOV_B32 84148480
+    %25:vgpr_32 = V_PERM_B32_e64 killed %22, killed %17, killed %24, implicit $exec
```
Remove the unrelated instructions

arsenm commented on llvm/test/CodeGen/AMDGPU/sdwa-peephole-vcnd_mask-2.mir:
```diff
@@ -0,0 +1,43 @@
+# changed to vcc_lo first. This would introduce a vcc_hi use that requires special
+# handling in si-peephole-sdwa.
+
+---
```
These different tests should go in the same test file

frederik-h replied:
They target incompatible ISAs, i.e. putting them in the same file crashes because of invalid register sizes.

arsenm replied:
you need to split wave32 vs. wave64, not by testcase

frederik-h replied:
I have tests for wave32 and wave64 in separate files now.

arsenm commented on llvm/test/CodeGen/AMDGPU/sdwa-peephole-vcnd_mask-2.mir:
```diff
@@ -0,0 +1,43 @@
+# RUN: llc %s -mtriple=amdgcn -mcpu=gfx1030 -run-pass=si-peephole-sdwa -o - | FileCheck %s
+
+# For conversion of_CNDMASK_B32_e64 to SDWA, the destination of V_CMP_O_F16_e64 must be
+# changed to vcc_lo first. This would introduce a vcc_hi use that requires special
+# handling in si-peephole-sdwa.
```
It's not really special, wave32 handling is just bad

frederik-h committed:
Change computeRegisterLiveness use

- Don' use auto
- Adapt other use for consistency
- Use default threshold
- Adjust tests

frederik-h committed:
Stop moving carry-in def instruction

frederik-h committed:
Handle undef carry-in operand

frederik-h committed:
Remove extra newline from debug output

frederik-h committed:
Rename test files to indicate the different ISAs being tested

frederik-h committed:
Use COPY instead of V_CMP_EQ for copy to VCC

frederik-h committed:
Handle wave32

frederik-h committed:
Rename sdwa-peephole-cndmask-gfx{9,10} tests

frederik-h committed:
Unify test names

arsenm commented on llvm/lib/Target/AMDGPU/SIPeepholeSDWA.cpp:
```diff
@@ -1061,6 +1063,61 @@ void SIPeepholeSDWA::pseudoOpConvertToVOP2(MachineInstr &MI,
+  if (CarryDef->isCompare() && TII->isVOP3(*CarryDef) &&
+      MRI->hasOneUse(CarryIn.getReg()))
+    CarryDef->substituteRegister(CarryIn.getReg(), Vcc, 0, *TRI);
+  else
+    BuildMI(MBB, MI, MI.getDebugLoc(), TII->get(AMDGPU::COPY))
+        .addReg(Vcc, RegState::Define)
+        .add(CarryIn);
```
```suggestion
  else {
    BuildMI(MBB, MI, MI.getDebugLoc(), TII->get(AMDGPU::COPY), Vcc)
        .add(CarryIn);
     }
```

arsenm commented on llvm/test/CodeGen/AMDGPU/sdwa-peephole-cndmask-vop3-wave32.mir:
```diff
@@ -0,0 +1,32 @@
+
+...
+---
+name: v_maximum_v2f16
```
Testname implies an unrelated purpose

arsenm commented on llvm/test/CodeGen/AMDGPU/sdwa-peephole-cndmask-vop2.mir:
```diff
@@ -0,0 +1,41 @@
+# NOTE: Assertions have been autogenerated by utils/update_mir_test_checks.py UTC_ARGS: --version 5
+# RUN: llc -mtriple=amdgcn-amd-amdhsa -run-pass=si-peephole-sdwa -mcpu=gfx803 -o - %s | FileCheck  -check-prefix=gfx8 %s
```
check prefixes should be all caps

arsenm commented on llvm/test/CodeGen/AMDGPU/sdwa-peephole-cndmask-vop3-wave64.mir:
```diff
@@ -0,0 +1,35 @@
+    $vgpr0 = COPY %5
+    SI_RETURN implicit $vgpr0
+
+...
```
Should have the one use and multi use tests. Also negative tests where source modifiers are used, and there's a live vcc use. Also the undef condition where getVRegDef would fail

arsenm commented on llvm/lib/Target/AMDGPU/SIPeepholeSDWA.cpp:
```diff
@@ -1061,6 +1063,61 @@ void SIPeepholeSDWA::pseudoOpConvertToVOP2(MachineInstr &MI,
+    return;
+  }
+
+  Register CarryReg = CarryIn.getReg();
+  MachineInstr *CarryDef = MRI->getVRegDef(CarryReg);
+  if (!CarryDef)
+    return;
```
Move above the computeRegisterLiveness

arsenm commented on llvm/lib/Target/AMDGPU/SIPeepholeSDWA.cpp:
```diff
@@ -1061,6 +1063,61 @@ void SIPeepholeSDWA::pseudoOpConvertToVOP2(MachineInstr &MI,
+/// copies from the carry-in to VCC.  The conversion will only be
+/// applied if \p MI can be shrunk to VOP2 and if VCC can be proven to
+/// be dead before \p MI.
+void SIPeepholeSDWA::convertToImplicitVcc(MachineInstr &MI,
```
Name is too broad, this is only about v_cndmask_b32_e64

frederik-h committed:
Update llvm/lib/Target/AMDGPU/SIPeepholeSDWA.cpp

Co-authored-by: Matt Arsenault <arsenm2@gmail.com>

frederik-h committed:
clang-format changes

frederik-h committed:
Rename convertToImplicitVcc and move CarryDef up

frederik-h committed:
Extend tests

Further changes:
- Add debug output for missing carry-in def.

frederik-h committed:
clang-format changes

frederik-h committed:
Change test prefix

arsenm commented on llvm/test/CodeGen/AMDGPU/sdwa-peephole-cndmask-vop3-wave64.mir:
```diff
@@ -0,0 +1,172 @@
+    %1:vgpr_32 = IMPLICIT_DEF
+    %2:vgpr_32 = V_PK_MAX_F16 8, undef %1, 8, undef %1, 0, 0, 0, 0, 0, implicit $mode, implicit $exec
+    %3:vgpr_32 = V_LSHRREV_B32_e64 16, undef %2, implicit $exec
+    %4:vgpr_32 = V_CNDMASK_B32_e64 0, 0, 0, %3, killed undef %0, implicit $exec
```
```suggestion
    %4:vgpr_32 = V_CNDMASK_B32_e64 0, 0, 0, %3, killed %0, implicit $exec
```

arsenm commented on llvm/test/CodeGen/AMDGPU/sdwa-peephole-cndmask-vop3-wave64.mir:
```diff
@@ -0,0 +1,172 @@
+    %1:vgpr_32 = IMPLICIT_DEF
+    %2:vgpr_32 = V_PK_MAX_F16 8, undef %1, 8, undef %1, 0, 0, 0, 0, 0, implicit $mode, implicit $exec
+    %3:vgpr_32 = V_LSHRREV_B32_e64 16, %2, implicit $exec
+    %4:vgpr_32 = V_CNDMASK_B32_e64 0, 0, 0, %3, killed undef %0, implicit $exec
```
```suggestion
    %4:vgpr_32 = V_CNDMASK_B32_e64 0, 0, 0, %3, killed %0, implicit $exec
```
Remove all of these undef flags for values that have actually have a def. There should only be one edge case test with the missing def

arsenm commented on llvm/test/CodeGen/AMDGPU/sdwa-peephole-cndmask-vop3-wave64.mir:
```diff
@@ -0,0 +1,172 @@
+    ; CHECK-NEXT: SI_RETURN implicit $vgpr0
+    %0:vgpr_32 = IMPLICIT_DEF
+    %2:vgpr_32 = IMPLICIT_DEF
+    %3:sreg_64_xexec = V_CMP_EQ_U32_e64 undef %0, 1, implicit $exec
```
```suggestion
    %3:sreg_64_xexec = V_CMP_EQ_U32_e64 %0, 1, implicit $exec
```

arsenm commented on llvm/test/CodeGen/AMDGPU/sdwa-peephole-cndmask-vop3-wave64.mir:
```diff
@@ -0,0 +1,172 @@
+    %0:vgpr_32 = IMPLICIT_DEF
+    %2:vgpr_32 = IMPLICIT_DEF
+    %3:sreg_64_xexec = V_CMP_EQ_U32_e64 undef %0, 1, implicit $exec
+    %5:vgpr_32 = V_LSHRREV_B32_e64 16, undef %2, implicit $exec
+    V_CMP_EQ_U32_e32 1, undef %2, implicit-def $vcc, implicit $exec
```
```suggestion
    %5:vgpr_32 = V_LSHRREV_B32_e64 16, %2, implicit $exec
    V_CMP_EQ_U32_e32 1, %2, implicit-def $vcc, implicit $exec
```

arsenm commented on llvm/test/CodeGen/AMDGPU/sdwa-peephole-cndmask-vop3-wave64.mir:
```diff
@@ -0,0 +1,172 @@
+    ; CHECK-NEXT: $vgpr0 = COPY [[V_CNDMASK_B32_e64_]]
+    ; CHECK-NEXT: SI_RETURN implicit $vgpr0
+    %1:vgpr_32 = IMPLICIT_DEF
+    %2:vgpr_32 = V_LSHRREV_B32_e64 16, undef %1, implicit $exec
+    %3:vgpr_32 = V_CNDMASK_B32_e64 0, 0, 0, %2, undef %0, implicit $exec
```
```suggestion
    %2:vgpr_32 = V_LSHRREV_B32_e64 16, %1, implicit $exec
    %3:vgpr_32 = V_CNDMASK_B32_e64 0, 0, 0, %2, undef %0, implicit $exec
```

frederik-h committed:
Apply suggestions from code review

Co-authored-by: Matt Arsenault <arsenm2@gmail.com>

frederik-h committed:
Adjusts tests

- Compact reg numbers in vop test
- Remove "undef"
- Readjust types in wave32 test

frederik-h committed:
Make sure that V_CND_MASK gets handled

frederik-h committed:
Change tests to avoid the impression that the carry-in def will be rewritten to VOPC

frederik-h commented on llvm/lib/Target/AMDGPU/SIPeepholeSDWA.cpp:
```diff
@@ -1061,6 +1063,62 @@ void SIPeepholeSDWA::pseudoOpConvertToVOP2(MachineInstr &MI,
+
+  // Change destination of compare instruction to VCC
+  // or copy to VCC if carry-in is not a compare inst.
+  if (CarryDef->isCompare() && TII->isVOP3(*CarryDef) &&
+      MRI->hasOneUse(CarryIn.getReg()))
```
My initial implementation was actually converting to the vop2 form. I removed this conversion in an attempt to simplify the code. I suppose the actual conversion of the carry-in def to vop2 is not really necessary here? I have changed the condition to make sure that v_cmp_class and all instructions which have a vopc form will be included.

frederik-h replied:
I am not sure if always using the copy has any negative effects on the generated code. I imagine it could be ok as well. I will give it a try.

arsenm replied:
My expectation is the physical register def in the instruction has a net worse effect than the separate copy

frederik-h replied:
I have changed the code to always use a copy now. The results are quite similar, except for e.g. this https://github.com/llvm/llvm-project/pull/137930/commits/135d3a09921d2382c6eb2553ea7533be49a93394#discussion_r2073100238.

frederik-h committed:
Always copy from carry-in operand to VCC

frederik-h commented on llvm/test/CodeGen/AMDGPU/fmin_legacy.f16.ll:
```diff
@@ -114,8 +114,10 @@ define <2 x half> @test_fmin_legacy_ule_v2f16(<2 x half> %a, <2 x half> %b) #0 {
 ; VI-SAFE-NEXT:    s_waitcnt vmcnt(0) expcnt(0) lgkmcnt(0)
-; VI-SAFE-NEXT:    v_cmp_ngt_f16 vcc, v0, v1 src0_sel:WORD_1 src1_sel:WORD_1
-; VI-SAFE-NEXT:    v_cndmask_b32_sdwa v2, v1, v0, vcc dst_sel:WORD_1 dst_unused:UNUSED_PAD src0_sel:WORD_1 src1_sel:WORD_1
+; VI-SAFE-NEXT:    v_lshrrev_b32_e32 v2, 16, v1
```
@arsenm This is an effect of always using the copy from the carry-in operand to VCC.

arsenm commented on llvm/test/CodeGen/AMDGPU/sdwa-peephole-cndmask-vop3-wave32.mir:
```diff
@@ -0,0 +1,201 @@
+    ; CHECK-NEXT: SI_RETURN implicit $vgpr0
+    %0:sreg_32_xm0_xexec = IMPLICIT_DEF
+    %1:vgpr_32 = IMPLICIT_DEF
+    %2:vgpr_32 = V_PK_MAX_F16 8, undef %1, 8, undef %1, 0, 0, 0, 0, 0, implicit $mode, implicit $exec
```
```suggestion
    %2:vgpr_32 = V_PK_MAX_F16 8, %1, 8, %1, 0, 0, 0, 0, 0, implicit $mode, implicit $exec
```

arsenm commented on llvm/test/CodeGen/AMDGPU/sdwa-peephole-cndmask-vop3-wave32.mir:
```diff
@@ -0,0 +1,201 @@
+    ; CHECK-NEXT: SI_RETURN implicit $vgpr0
+    %0:sreg_32_xm0_xexec = IMPLICIT_DEF
+    %1:vgpr_32 = IMPLICIT_DEF
+    %2:vgpr_32 = V_PK_MAX_F16 8, undef %1, 8, undef %1, 0, 0, 0, 0, 0, implicit $mode, implicit $exec
```
```suggestion
    %2:vgpr_32 = V_PK_MAX_F16 8, %1, 8, %1, 0, 0, 0, 0, 0, implicit $mode, implicit $exec
```

arsenm commented on llvm/test/CodeGen/AMDGPU/sdwa-peephole-cndmask-vop3-wave32.mir:
```diff
@@ -0,0 +1,201 @@
+    ; CHECK-NEXT: SI_RETURN implicit $vgpr0
+    %0:sreg_32_xm0_xexec = IMPLICIT_DEF
+    %2:vgpr_32 = IMPLICIT_DEF
+    %3:sreg_32_xm0_xexec = V_CMP_EQ_U32_e64 undef %0, 1, implicit $exec
+    %5:vgpr_32 = V_LSHRREV_B32_e64 16, undef %2, implicit $exec
```
```suggestion
    %3:sreg_32_xm0_xexec = V_CMP_EQ_U32_e64 %0, 1, implicit $exec
    %5:vgpr_32 = V_LSHRREV_B32_e64 16, %2, implicit $exec
```

arsenm commented on llvm/test/CodeGen/AMDGPU/sdwa-peephole-cndmask-vop3-wave32.mir:
```diff
@@ -0,0 +1,201 @@
+    ; CHECK-NEXT: SI_RETURN implicit $vgpr0
+    %0:sreg_32_xm0_xexec = IMPLICIT_DEF
+    %1:vgpr_32 = IMPLICIT_DEF
+    %2:sreg_32_xm0_xexec = V_CMP_EQ_U32_e64 undef %0, 0, implicit $exec
+    %3:vgpr_32 = V_LSHRREV_B32_e64 16, undef %1, implicit $exec
```
```suggestion
    %2:sreg_32_xm0_xexec = V_CMP_EQ_U32_e64 %0, 0, implicit $exec
    %3:vgpr_32 = V_LSHRREV_B32_e64 16, %1, implicit $exec
```

arsenm commented on llvm/test/CodeGen/AMDGPU/sdwa-peephole-cndmask-vop3-wave32.mir:
```diff
@@ -0,0 +1,201 @@
+    ; CHECK-NEXT: $vgpr0 = COPY [[V_CNDMASK_B32_e64_]]
+    ; CHECK-NEXT: SI_RETURN implicit $vgpr0
+    %1:sreg_32_xm0_xexec = IMPLICIT_DEF
+    %2:vgpr_32 = V_LSHRREV_B32_e64 16, undef %1:sreg_32_xm0_xexec, implicit $exec
```
```suggestion
    %2:vgpr_32 = V_LSHRREV_B32_e64 16, %1:sreg_32_xm0_xexec, implicit $exec
```

arsenm commented on llvm/test/CodeGen/AMDGPU/sdwa-peephole-cndmask-vop3-wave64.mir:
```diff
@@ -0,0 +1,201 @@
+    ; CHECK-NEXT: SI_RETURN implicit $vgpr0
+    %0:vgpr_32 = IMPLICIT_DEF
+    %1:vgpr_32 = IMPLICIT_DEF
+    %2:sreg_64_xexec = V_CMP_EQ_U32_e64 undef %0, 1, implicit $exec
+    %3:vgpr_32 = V_LSHRREV_B32_e64 16, undef %1, implicit $exec
```
```suggestion
    %2:sreg_64_xexec = V_CMP_EQ_U32_e64 %0, 1, implicit $exec
    %3:vgpr_32 = V_LSHRREV_B32_e64 16, %1, implicit $exec
```

arsenm commented on llvm/test/CodeGen/AMDGPU/sdwa-peephole-cndmask-vop3-wave64.mir:
```diff
@@ -0,0 +1,201 @@
+    ; CHECK-NEXT: SI_RETURN implicit $vgpr0
+    %0:sreg_32_xm0_xexec = IMPLICIT_DEF
+    %1:vgpr_32 = IMPLICIT_DEF
+    %2:sreg_64_xexec = V_CMP_CLASS_F32_e64 2, undef %0, 1, implicit $exec
+    %3:vgpr_32 = V_LSHRREV_B32_e64 16, undef %1, implicit $exec
```
```suggestion
    %2:sreg_64_xexec = V_CMP_CLASS_F32_e64 2,%0, 1, implicit $exec
    %3:vgpr_32 = V_LSHRREV_B32_e64 16, %1, implicit $exec
```

arsenm reviewed (commented):
lgtm except there are still tests using undef flags on operands which really have defs

frederik-h committed:
Apply suggestions from code review

Co-authored-by: Matt Arsenault <arsenm2@gmail.com>

frederik-h committed:
Apply suggestions from code review

Co-authored-by: Matt Arsenault <arsenm2@gmail.com>

arsenm commented on llvm/test/CodeGen/AMDGPU/sdwa-peephole-cndmask-vop2.mir:
```diff
@@ -0,0 +1,41 @@
+# RUN: llc -mtriple=amdgcn-amd-amdhsa -run-pass=si-peephole-sdwa -mcpu=gfx803 -o - %s | FileCheck  -check-prefix=GFX8 %s
+# RUN: llc -mtriple=amdgcn-amd-amdhsa -run-pass=si-peephole-sdwa -mcpu=gfx1100 -o - %s | FileCheck -check-prefix=GFX11 %s
+
+---
```
This test should merge into the other 2. This also should really have a separate wave32 and wave64 version, and it's a bug you need to use vcc in wave32

frederik-h replied:
> This test should merge into the other 2. This also should really have a separate wave32 and wave64 version, and it's a bug you need to use vcc in wave32

Ok, I will merge the tests. Concerning the bug: You mean having to specify `implicit $vcc` in ` %4:vgpr_32 = V_CNDMASK_B32_e32 killed %2, killed %3, implicit $exec, implicit $vcc` with "wave32"?

arsenm replied:
It should be vcc_lo. Ideally the instruction definitions would be swapped out depending on HwMode

frederik-h replied:
> This test should merge into the other 2. 
Done ...

> This also should really have a separate wave32 and wave64 version, and it's a bug you need to use vcc in wave32
 ... and this is an unrelated/follow-up issue if I understand this correctly.

frederik-h committed:
Adjust tests

- Move VOP2 tests cases into VOP3 wave32, wave64 test files
- Adjust test names and comments to the fact that we no longer attempt
to change the carry-in operands to VOP2.

frederik-h committed:
Rename VOP2 test file and remove "-vop3" from other test names

... since those files now also cover the vop2 case.

frederik-h commented:
@arsenm Thanks for the very helpful review!
